x/gamm/pool-models/balancer/pool_test.go

```diff
@@ -0,0 +1,1285 @@
+	testcases := []testcase{
+		{
+			initialPoolOut:   1_000_000_000_000,
+			tokenOut:         100,
+			initialWeightOut: 100,
+			initialWeightIn:  100,
+		},
+		{
+			initialPoolOut:   1_000_000_000_000,
+			tokenOut:         100,
+			initialWeightOut: 50,
+			initialWeightIn:  100,
+		},
+		{
+			initialPoolOut:   1_000_000_000_000,
+			tokenOut:         50,
+			initialWeightOut: 100,
+			initialWeightIn:  100,
+		},
+		{
+			initialPoolOut:   1_000_000_000_000,
+			tokenOut:         100,
+			initialWeightOut: 100,
+			initialWeightIn:  50,
+		},
+		{
+			initialPoolOut:   1_000_000,
+			tokenOut:         100,
+			initialWeightOut: 100,
+			initialWeightIn:  100,
+		},
+		{
+			initialPoolOut:   2_351_333,
+			tokenOut:         7,
+			initialWeightOut: 148,
+			initialWeightIn:  57,
+		},
+		{
+			initialPoolOut:   1_000,
+			tokenOut:         25,
+			initialWeightOut: 100,
+			initialWeightIn:  100,
+		},
+		{
+			initialPoolOut:   1_000,
+			tokenOut:         26,
+			initialWeightOut: 100,
+			initialWeightIn:  100,
+		},
+	}
+
+	swapFeeCases := []string{"0", "0.001", "0.1", "0.5", "0.99"}
+
+	getTestCaseName := func(tc testcase, swapFeeCase string) string {
+		return fmt.Sprintf("initialPoolOut: %d, initialCalcOut: %d, initialWeightOut: %d, initialWeightIn: %d, swapFee: %s",
+			tc.initialPoolOut,
+			tc.tokenOut,
+			tc.initialWeightOut,
+			tc.initialWeightIn,
+			swapFeeCase,
+		)
+	}
+
+	for _, tc := range testcases {
+		for _, swapFee := range swapFeeCases {
+			t.Run(getTestCaseName(tc, swapFee), func(t *testing.T) {
+				swapFeeDec, err := sdk.NewDecFromStr(swapFee)
+				require.NoError(t, err)
+
+				initialPoolBalanceOut := sdk.NewInt(tc.initialPoolOut)
+
+				initialWeightOut := sdk.NewInt(tc.initialWeightOut)
+				initialWeightIn := sdk.NewInt(tc.initialWeightIn)
+
+				initialTotalShares := types.InitPoolSharesSupply.ToDec()
+				initialCalcTokenOut := sdk.NewInt(tc.tokenOut)
+
+				actualSharesOut := balancer.CalcPoolSharesOutGivenSingleAssetIn(
+					initialPoolBalanceOut.ToDec(),
+					initialWeightOut.ToDec().Quo(initialWeightOut.Add(initialWeightIn).ToDec()),
+					initialTotalShares,
+					initialCalcTokenOut.ToDec(),
+					swapFeeDec,
+				)
+
+				inverseCalcTokenOut := balancer.CalcSingleAssetInGivenPoolSharesOut(
+					initialPoolBalanceOut.Add(initialCalcTokenOut).ToDec(),
+					initialWeightOut.ToDec().Quo(initialWeightOut.Add(initialWeightIn).ToDec()),
+					initialTotalShares.Add(actualSharesOut),
+					actualSharesOut,
+					swapFeeDec,
+				)
+
+				tol := sdk.NewDec(1)
+				require.True(osmoutils.DecApproxEq(t, initialCalcTokenOut.ToDec(), inverseCalcTokenOut, tol))
+			})
+		}
+	}
+}
+
+// Expected is un-scaled
+func testTotalWeight(t *testing.T, expected sdk.Int, pool balancer.Pool) {
+	scaledExpected := expected.MulRaw(balancer.GuaranteedWeightPrecision)
+	require.Equal(t,
+		scaledExpected.String(),
+		pool.GetTotalWeight().String())
+}
+
+func TestBalancerPoolParams(t *testing.T) {
```

Comment:
We could move this to `pool_params_test.go` since params are now in its own file